### PR TITLE
Dev/1.17 entity origin

### PR DESCRIPTION
**Pull request by electronicboy**
Replaces #5793 for 1.17

Closes #5793.

### CODE REVIEW

File: patches/server/0025-Entity-Origin-API.patch [hunk 1]
```diff
@@ -17,66 +17,83 @@ index 88bac72edf19c578902f49d20353989ed4d96f8f..e79faeb26d079de0108268fd2607cf9e
...
 --- a/src/main/java/net/minecraft/world/entity/Entity.java
 +++ b/src/main/java/net/minecraft/world/entity/Entity.java
-@@ -280,6 +280,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+@@ -280,6 +280,23 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
      public org.bukkit.projectiles.ProjectileSource projectileSource; // For projectiles only
      public boolean forceExplosionKnockback; // SPIGOT-949
      public boolean persistentInvisibility = false;
-+    public org.bukkit.Location origin; // Paper
++    // Paper start
++    @javax.annotation.Nullable // Paper
```
Comment (Proximyst) [inline patches/server/0025-Entity-Origin-API.patch]: Paper start implies this is done by Paper :p

File: patches/server/0025-Entity-Origin-API.patch [hunk 2]
```diff
@@ -17,66 +17,83 @@ index 88bac72edf19c578902f49d20353989ed4d96f8f..e79faeb26d079de0108268fd2607cf9e
...
++    }
++
++    public org.bukkit.util.Vector getOriginVector() {
++        return this.origin != null ? this.origin.clone() : null;
++    }
++
++    public UUID getOriginWorld() {
++        return this.originWorld;
++    }
++    // Paper
```
Comment (Proximyst) [inline patches/server/0025-Entity-Origin-API.patch]:
```suggestion
+    // Paper end
```

File: patches/server/0025-Entity-Origin-API.patch [hunk 3]
```diff
@@ -17,66 +17,83 @@ index 88bac72edf19c578902f49d20353989ed4d96f8f..e79faeb26d079de0108268fd2607cf9e
...
++    // Paper start
++    @javax.annotation.Nullable // Paper
++    private org.bukkit.util.Vector origin; // Paper
++    private UUID originWorld; // Paper
++    public void setOrigin(Location location) {
++        this.origin = location.toVector();
++        this.originWorld = location.getWorld().getUID();
++    }
++
++    public org.bukkit.util.Vector getOriginVector() {
```
Comment (Proximyst) [inline patches/server/0025-Entity-Origin-API.patch:55]: Let's throw that nullable annotation here too

**Review by Proximyst — APPROVED** [review]
LGTM!

**Review by jpenilla — APPROVED** [review]

**Pull request merged**